Model/Sync/Catalog/Data.php

```diff
@@ -369,65 +378,69 @@ protected function prepareOptions($syncItems, $configIds, $productObjects)
      */
     public function attributeMapping(Product $item)
     {
-        $itemArray = [];
-        $mapAttributes = $this->mappingAttributes;
+        $itemAttributesData = [];
 
-        foreach ($mapAttributes as $key => $attr) {
+        foreach ($this->attributesMapping as $attributeKey => $attributeDetails) {
             try {
-                if ($key === 'gtins') {
-                    $value = $this->prepareGtinsData($attr, $item);
-                } elseif ($key === 'custom_properties') {
-                    $value = $this->prepareCustomProperties($attr, $item);
-                } elseif ($key === 'is_discontinued') {
-                    $value = false;
-                } else {
-                    if ($attr['default']) {
-                        $data = $item->getData($attr['attr_code']);
-
-                        if (isset($attr['type']) && $attr['type'] === 'url') {
-                            $data = $item->getProductUrl();
+                switch ($attributeKey) {
```

Comment:
```suggestion
                $attributeDataValue = '';
                $attributeDetailsAttributeCode = '';
                $attributeDetailsMethod = $attributeDetails['method'];
                if (isset($attributeDetails['attr_code']) && $attributeDetails['attr_code']) {
                    $attributeDetailsAttributeCode = $attributeDetails['attr_code'];
                }
                $itemValue = $this->$attributeDetailsMethod($item, $attributeDetailsAttributeCode);

                switch ($attributeKey) {
                    case 'gtins':
                        $attributeDataValue = $this->prepareGtinsData($attributeDetails, $item);
                        break;
                    case 'custom_properties':
                        $attributeDataValue = $this->prepareCustomProperties($attributeDetails, $item);
                        break;
                    case 'is_discontinued':
                        $attributeDataValue = false;
                        break;
                    case 'url':
                        $attributeDataValue = $item->getProductUrl();
                        break;
                    case 'image_url':
                        $attributeDataValue = $this->getProductImageUrl($item);
                        break;
                    case 'group_name':
                        if ($itemValue) {
                            $attributeDataValue = substr((string)$itemValue, 0, 100);
                            $attributeDataValue = strtolower($attributeDataValue);
                            $attributeDataValue = str_replace(' ', '_', $attributeDataValue);
                            $attributeDataValue = preg_replace('/[^A-Za-z0-9_-]/', '-', $attributeDataValue);
                        }
                        break;
                    default:
                        if (!$attributeDetails['default'] && isset($attributeDetails['method']) && $attributeDetails['method']) {
                            if ($itemValue) {
                                $attributeDataValue = $itemValue;
                            } elseif ($attributeDetailsMethod == 'getProductPrice') {
                                $attributeDataValue = 0.00;
                            } else {
                                $attributeDataValue = $itemValue;
                            }
                        }
                }

                $itemAttributesData[$attributeKey] = $attributeDataValue;
                if (($attributeKey == 'custom_properties' || $attributeKey == 'gtins') && !$attributeDataValue) {
                    unset($itemAttributesData[$attributeKey]);
                }
```

Reply:
I don't agree with this change. You've basically extracted what we had in the `default` clause, but it's not relevant for all of the action types.
It looks better for sure, but it degregates performance.

Reply:
I disagree with that. Because default clause shouldn't check for other cases. Its a bad practice.
But I agree on the performance concern. I edited the suggestionso it will be better performance wise. Let me know what you think. We can also talk about it in a call.

Reply:
I disagree with that. Because default clause shouldn't check for other cases. Its a bad practice.
But I agree on the performance concern. I created a new branch with a refactor commit so it will be better performance wise (and easier for you to read than a suggestion). Let me know what you think. We can also talk about it in a call.
https://github.com/YotpoLtd/magento2-module-core/pull/197/commits/7f554dc20345b8d292443e0ac96884d8e1b38063